src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -392,6 +434,37 @@ pub struct VcpuState {
     vcpu_events: kvm_vcpu_events,
     xcrs: kvm_xcrs,
     xsave: kvm_xsave,
+    #[version(start = 2, default_fn = "default_tsc_khz", ser_fn = "ser_tsc")]
+    pub tsc_khz: Option<u32>,
+}
+
+impl VcpuState {
+    fn default_tsc_khz(_: u16) -> Option<u32> {
+        warn!("CPU TSC freq not found in snapshot");
+        None
+    }
+
+    fn ser_tsc(&mut self, target_version: u16) -> VersionizeResult<()> {
+        if target_version < 2 {
```

Comment:
This can be simplified, no need to check the target_version here, please see https://docs.rs/versionize_derive/0.1.4/versionize_derive/derive.Versionize.html#ser_fn

Also I would add asserts in tests for the 2 warnings.

Reply:
I refactored this and added a negative test for the second message. The first one is really tough to get. I left a TODO there for when we decide to have mock structures for testing here, so we can easily cover the case when `get_tsc_khz` doesn't work.